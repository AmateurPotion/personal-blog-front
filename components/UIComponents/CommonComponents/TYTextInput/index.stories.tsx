/**
 * After storybook v5.3, They recommend use CSF(Component Story Format)
 * instead of using storiesof API
 */
import React from "react";
import Component from "./index";
import { withKnobs, text, number } from "@storybook/addon-knobs";
import { useCallback, useState } from "@storybook/addons";
import { action } from "@storybook/addon-actions";

export default {
    title: "COMPONENTS|TY/TYTextInput",
    component: Component,
    decorators: [withKnobs]
};

export const standard = () => {
    const [value, setValue] = useState("");
    const placeholder = text("placeholder", "placeholder");
    const width = text("width", "200px");
    const limit = number("limit", 10);

    const onChange = useCallback(
        (event: any) => {
            setValue(event.target.value);
            action("onChange value");
        },
        [action]
    );
    return <Component value={value} width={width} limit={limit} onChange={onChange} placeholder={placeholder} />;
};

export const actionTest = () => {
    const [value, setValue] = useState("");
    const placeholder = text("placeholder", "placeholder");
    return <Component value={value} onChange={action("onChange")} placeholder={placeholder} />;
};
