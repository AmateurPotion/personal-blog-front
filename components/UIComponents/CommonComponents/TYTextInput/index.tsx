import React, { useRef, useCallback, useState } from "react";
import styled from "styled-components";

interface IProps {
    type: "text" | "password";
    value: any;
    onChange: (eveny: any) => void;
    label?: string;
    placeholder?: string;
    limit?: number;
    width?: string;
}

const TYTextInput: React.FC<IProps> = props => {
    const { type, value, onChange, placeholder, limit, width, label } = props;
    const [isFocus, setIsFocus] = useState(false);
    const inputRef = useRef<HTMLInputElement>(null);

    const _onChange = useCallback(
        (event: any) => {
            const shouldLimit = !!!limit || event.target.value.length <= limit;
            if (shouldLimit) {
                onChange(event);
            }
        },
        [onChange, limit]
    );

    const _focusInput = useCallback(() => {
        setIsFocus(true);
    }, []);
    const _blurInput = useCallback(() => {
        setIsFocus(false);
    }, []);

    return (
        <>
            <Wrapper>
                <Label>{label}</Label>
                <Container width={width}>
                    <Input
                        type={type}
                        ref={inputRef}
                        value={value}
                        placeholder={placeholder}
                        onChange={_onChange}
                        onFocus={_focusInput}
                        onBlur={_blurInput}
                    />
                </Container>
            </Wrapper>
        </>
    );
};

TYTextInput.defaultProps = {
    onChange: () => {},
    value: "",
    label: "",
    placeholder: "placeholder",
    width: "200px"
};

const Wrapper = styled.div`
    margin: 1rem;
`;

const Label = styled.label`
    display: block;
    margin-bottom: 0.6rem;
`;

const Container = styled.div<{ width: string }>`
    display: flex;
    position: relative;
    width: ${props => props.width};
`;

const Input = styled.input`
    padding: 0.8rem 1rem;
    width: 100%;
    border: none;
    border: 1px solid ${props => props.theme.achromaticColor};
    border-radius: 0.2rem;
    font-weight: 700;
    color: "black";
    text-align: left;
    &:hover {
        border: 1px solid ${props => props.theme.subColor};
    }
    &:focus {
        border: 1px solid ${props => props.theme.mainColor};
        outline: none;
    }
    &::placeholder {
        color: ${props => props.theme.achromaticColor};
    }
`;

export default TYTextInput;
