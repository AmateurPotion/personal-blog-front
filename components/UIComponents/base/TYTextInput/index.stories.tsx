/**
 * After storybook v5.3, They recommend use CSF(Component Story Format)
 * instead of using storiesof API
 */
import React from "react";
import Component from "./index";
import { withKnobs, text, number, select, color } from "@storybook/addon-knobs";
import { useCallback, useState } from "@storybook/addons";
import { action } from "@storybook/addon-actions";

export default {
    title: "COMPONENTS|TY/TYTextInput",
    component: Component,
    decorators: [withKnobs]
};

export const standard = () => {
    const [value, setValue] = useState("");

    const options = {
        text: "text",
        password: "password"
    };

    const defaultValue = options.text;

    const type = select("type", options, defaultValue, "TY-TEXTINPUT-01");
    const placeholder = text("placeholder", "12자 이상의 아이디를 입력하세요", "TY-TEXTINPUT-01");
    const label = text("label", "아이디", "TY-TEXTINPUT-01");
    const width = text("width", "200px", "TY-TEXTINPUT-01");
    const limit = number("limit", 10, {}, "TY-TEXTINPUT-01");
    const labelColor = color("labelColor", "", "TY-TEXTINPUT-01");
    const outlineColor = color("outlineColor", "", "TY-TEXTINPUT-01");

    const onChange = useCallback(
        (event: any) => {
            setValue(event.target.value);
            action("onChange value");
        },
        [action]
    );

    return (
        <Component
            type={type as any}
            label={label}
            value={value}
            width={width}
            limit={limit}
            onChange={onChange}
            placeholder={placeholder}
            labelColor={labelColor}
            outLineColor={outlineColor}
        />
    );
};

export const actionTest = () => {
    const [value, setValue] = useState("");
    const placeholder = text("placeholder", "아이디");

    return <Component type={"text"} value={value} onChange={action("onChange")} placeholder={placeholder} />;
};
