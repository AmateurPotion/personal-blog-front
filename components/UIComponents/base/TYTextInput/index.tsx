import React, { useRef, useCallback, useState, ChangeEvent } from "react";
import styled from "styled-components";

interface IProps {
    type: "text" | "password";
    value: any;
    onChange: (event: ChangeEvent<HTMLInputElement>) => void;
    label?: string;
    placeholder?: string;
    limit?: number;
    width?: string;
    outLineColor?: string;
    labelColor?: string;
    labelType?: "top" | "left";
}

const TYTextInput: React.FC<IProps> = props => {
    const { type, value, onChange, placeholder, limit, width, label, labelColor, labelType } = props;
    const [isFocus, setIsFocus] = useState(false);
    const inputRef = useRef<HTMLInputElement>(null);

    const _onChange = useCallback(
        (event: ChangeEvent<HTMLInputElement>) => {
            const shouldLimit = !!!limit || event.target.value.length <= limit;
            if (shouldLimit) {
                onChange(event);
            }
        },
        [onChange, limit]
    );

    const _focusInput = useCallback(() => {
        setIsFocus(true);
    }, []);
    const _blurInput = useCallback(() => {
        setIsFocus(false);
    }, []);

    return (
        <>
            <Wrapper {...props}>
                <Container width={width} limit={limit}>
                    <Label labelColor={labelColor} labelType={labelType}>
                        {label}
                    </Label>
                    <Input
                        type={type}
                        ref={inputRef}
                        value={value}
                        placeholder={placeholder}
                        onChange={_onChange}
                        onFocus={_focusInput}
                        onBlur={_blurInput}
                        {...props}
                    />
                </Container>
            </Wrapper>
        </>
    );
};

TYTextInput.defaultProps = {
    onChange: () => {},
    value: "",
    label: "",
    placeholder: "placeholder",
    width: "200px",
    outLineColor: "#e03131",
    labelColor: "#000000",
    labelType: "top"
};

const Wrapper = styled.div<Partial<IProps>>`
    margin: 1rem;
    @media screen and (max-width: ${props => props.theme.mediumPoint}) {
        margin: 1rem auto;
        width: 96%;
    }
`;

const Container = styled.div<Partial<IProps> & { width: string }>`
    display: block;
    position: relative;
    width: ${props => props.width};
    @media screen and (max-width: ${props => props.theme.mediumPoint}) {
        width: ${props => props.limit * 3}%;
        margin: 0 auto;
    }
`;

const Label = styled.label<Partial<IProps>>`
    display: block;
    font-weight: 900;
    text-align: left;
    margin-bottom: 0.6rem;
    color: ${props => props.labelColor};
`;

const Input = styled.input<Partial<IProps>>`
    display: inline-block;
    padding: 0.8rem 1rem;
    width: 90%;
    border: 1px solid #dee2e6;
    border-radius: 0.2rem;
    font-weight: 700;
    color: black;
    text-align: left;
    &:hover {
        border: 1px solid ${props => props.outLineColor};
    }
    &:focus {
        border: 1px solid ${props => props.outLineColor};
        outline: none;
    }
    &::placeholder {
        color: #ced4da;
    }
`;

export default TYTextInput;
