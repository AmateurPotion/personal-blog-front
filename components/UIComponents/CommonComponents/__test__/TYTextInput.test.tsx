import React, {useState, useCallback} from "react";
import TYTextInput from "../TYTextInput";
import renderer from "react-test-renderer";
import { mount } from "enzyme";

describe("[TYTextInput] Component Rendering", () => {
    it("should render without crashing", () => {
        const value = "";
        const snapshot = renderer.create(<TYTextInput value={value} onChange={()=>{}} placeholder={"placeholder"} />).toJSON();
        expect(snapshot).toMatchSnapshot();
    });

    it("render placholder without crashing", () => {
        const value = "";
        const wrapper = mount(<TYTextInput value={value} onChange={()=>{}} placeholder={"test-placeholder"} />);
        const input = wrapper.find("input");
        expect(wrapper.props().placeholder).toBe("test-placeholder");
    });

    it("has an input element", () => {
        const value = ""
        const wrapper = mount(<TYTextInput value={value} onChange={()=>{}} />);
        const inputElement = wrapper.find("input[type='text']");
        expect(inputElement.exists()).toBe(true);
    });

    it("simulate input change", () => {
        let value = ""
        const mockedEvent = {
            target : {
                value: "It is test data"
            }
        }
        const wrapper = mount(<TYTextInput value={value} onChange={(event) => {
            value = event.target.value;
        }} />);
        let inputElement = wrapper.find("input[type='text']");
        inputElement.simulate('change', mockedEvent);

        /**
        * Document for me
        * Like const [value, setValue] = useState("");
        * After changing state from here, set value prop again 
        * and check input element's value from here
        */
        wrapper.setProps({ value: value });
        inputElement = wrapper.find("input[type='text']");

        expect(inputElement.props().value).toBe("It is test data");
    });
});
