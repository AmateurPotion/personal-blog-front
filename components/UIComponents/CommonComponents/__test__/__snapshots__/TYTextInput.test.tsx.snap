// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[TYTextInput] Component Rendering should render without crashing 1`] = `
<div
  className="sc-bdVaJa HISMw"
  width="200px"
>
  <input
    className="sc-bwzfXH dVJTUB"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    placeholder="placeholder"
    type="text"
    value=""
  />
  <div
    className="sc-htpNat gMiaFE"
  />
</div>
`;
